modules/utils/2.0/utils.smk

```diff
@@ -0,0 +1,149 @@
+#!/usr/bin/env snakemake
+
+
+##### ATTRIBUTION #####
+
+
+# Original Snakefile Author:    Helena Winata, Bruno Grande
+# Module Author:                Bruno Grande
+# Additional Contributors:      N/A
+
+
+##### SETUP #####
+
+import sys, os
+from os.path import join
+
+# Import package with useful functions for developing analysis modules
+import oncopipe as op
+
+_UTILS = config["lcr-modules"]["utils"]
+LOG = "/logs/" + op._session.launched_fmt
+
+wildcard_constraints:
+    prefix = "[0-9]{2}-.*"
+
+
+##### RULES #####
+# _utils_bam_sort: Sort a BAM file using coordinates
+rule:
+    input:
+        bam = "{out_dir}/{prefix}/{suffix}.bam"
+    output:
+        bam = "{out_dir}/{prefix}/{suffix}.sort.bam"
+    log:
+        stdout = "{out_dir}" + LOG + "/{prefix}/{suffix}_sort.stdout.log",
+        stderr = "{out_dir}" + LOG + "/{prefix}/{suffix}_sort.stderr.log"
+    wildcard_constraints:
+        prefix = ".*(sort).*"
+    params:
+        #logs = _get_log_dirs,
+        opts = _UTILS["options"]["bam_sort"],
+        prefix ="{out_dir}/{prefix}/{suffix}"
+    conda:
+        _UTILS["conda_envs"]["samtools"]
+    threads:
+        _UTILS["threads"]["bam_sort"]
+    resources: 
+        mem_mb = _UTILS["mem_mb"]["bam_sort"]
+    shell:
+        op.as_one_line("""
+        samtools sort 
+        -@ {threads} -m $(({resources.mem_mb} / {threads}))M
+        {params.opts}
+        -T {params.prefix} -o {output.bam} 
+        {input.bam} 
+        > {log.stdout}
+        2> {log.stderr}
+        """)
+
+
+# _utils_bam_markdups: Mark duplicates in a BAM file using Picard criteria
+rule:
+    input:
+        bam = "{out_dir}/{prefix}/{suffix}.bam"
+    output:
+        bam = "{out_dir}/{prefix}/{suffix}.mdups.bam"
+    log:
+        stdout = "{out_dir}" + LOG + "/{prefix}/{suffix}_mark_dups.stdout.log",
+        stderr = "{out_dir}" + LOG + "/{prefix}/{suffix}_mark_dups.stderr.log"
+    wildcard_constraints:
+        prefix = ".*(mark_dups).*"
```

Comment:
Maybe I don't understand the purpose of this constraint, but should it be "mdups" instead of "mark_dups" since the output name has the ".mdups" addition?

Reply:
I think it'll make more sense if you review this part with the `bwa_mem` module. it's because the subdirectory is named `02-mark_dups`